Add initial external assets adjuncts work

JackLewis-digirati opened the pull request:
Resolves #1073

Implements the `/adjuncts` endpoint for external assets.

This is a fairly standard `CRUD` controller that sits under `asset` and currently can only be seen with the `API`.

Additionally, `adjuncts` has been added as a link in an `image`, following details from here - https://deploy-preview-2--dlcs-docs.netlify.app/api-doc/registering-assets

> [!NOTE]
> Also updates appetiser in the local docker compose file

> [!NOTE]
> Contains a migration

JackLewis-digirati committed:
Add initial adjuncts work

- Creates hydra + DB models
- Adds a migration to the DB
- Creates hydra/model converters
- Controller created with placeholders for all endpoints
- Initial implementations of GET and POST

stephenwf commented on src/protagonist/DLCS.Repository/Migrations/DlcsContextModelSnapshot.cs:
```diff
@@ -30,6 +30,53 @@ protected override void BuildModel(ModelBuilder modelBuilder)
+                    b.Property<string>("IiifLink")
+                        .IsRequired()
+                        .HasMaxLength(500)
+                        .HasColumnType("character varying(500)");
```
Because it would require a DB change later, I think it's best to mention now even if it is not implemented at the moment. I think this should be optional. I think there would be a benefit in being able to add an intermediate step that _doesn't_ appear in the final IIIF - or is supplemental to it.

donaldgray replied:
That functionality will be covered later. Currently IIIFLink can be `seeAlso`, `annotations` or `rendering`.

In a future ticket we'll handle IIIFLink being empty - in that case it will be output on any generated Manifests in an `"otherAdjuncts"` property (see https://deploy-preview-2--dlcs-docs.netlify.app/api-doc/adjuncts#iiiflink). The idea is that we are advertising it but not in a proper IIIF compliant manner. 

When NQ is generated, the consumer can do with it what they will - ignore it, move it etc. They important thing is they'll know it exists and will have paths to request it etc.

JackLewis-digirati committed:
Allow PUT to work

- Adds a created/updated to Adjunct
- Modifies IiifLink to be IIIFLink
- modifies CreateOrUpdateAdjunct to work with PUT

JackLewis-digirati committed:
Enable Delete + GetAll adjuncts to work

- Creates DELETE endpoint
- makes sure DELETE asset cascades to adjuncts
- Creates paged fetch endpoint
- Adds tests for location
- Enables @type to work across POST/PUT
- Adds a ResourceType class to be used for @Type

JackLewis-digirati committed:
Final steps before review

- Remove IIIF from the HydraModel class, along with using a Dictionary
- set ExternalId to required
- Create a dictionary to LanguageMap converter
- Remove adjuncts from asset notification sender
- Stop validating type value
- Revert ResourceType.cs

JackLewis-digirati committed:
Add migration

JackLewis-digirati committed:
remove unneeded $

JackLewis-digirati committed:
Update appetiser compose

JackLewis-digirati committed:
Add adjuncts link to asset

JackLewis-digirati committed:
Modify properties

- Adds finished and publicId
- Removes modified
- Updates migration

JackLewis-digirati committed:
Remove paging

donaldgray commented on src/protagonist/API/Features/Adjuncts/AdjunctsController.cs:
```diff
@@ -0,0 +1,131 @@
+        return await CreateOrUpdateAdjunct(customerId, spaceId, imageId, hydraAdjunct, validator, false, cancellationToken);
+    }
+    
+    private async Task<IActionResult> CreateOrUpdateAdjunct(int customerId, int spaceId, string imageId, Adjunct hydraAdjunct,
```
Small, consistency - other classes in Protagonist have `public` methods first/higher in class, with `private`/`protected` etc lower. 

This can be particularly useful for Controllers as these are the API endpoints

donaldgray commented on src/protagonist/API/Features/Adjuncts/AdjunctsController.cs:
```diff
@@ -0,0 +1,131 @@
+    /// </summary>
+    /// <returns>A delete result.</returns>
+    [HttpDelete("{adjunctId}")]
+    [ProducesResponseType(200, Type = typeof(HydraCollection<Adjunct>))]
```
Will return 204|No Content rather than 200/list of Adjuncts

donaldgray commented on src/protagonist/API/Features/Adjuncts/AdjunctsController.cs:
```diff
@@ -0,0 +1,131 @@
+    /// <summary>
+    /// Get details of all adjuncts for an asset.
+    /// </summary>
+    /// <returns>A Hydra JSON-LD Adjunct object representing the adjuncts.</returns>
```
Small - returns a Hydra Collection, rather than an Adjunct (same with PUT/POST)

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/GetAdjuncts.cs:
```diff
@@ -0,0 +1,29 @@
+public class GetAdjuncts(AssetId assetId) : IRequest<FetchEntityResult<IReadOnlyCollection<Adjunct>>>, IPagedRequest
+{
+    public int Page { get; set; }
+    public int PageSize { get; set; }
```
This no longer needs to be an `IPagedRequest`

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/GetAdjuncts.cs:
```diff
@@ -0,0 +1,29 @@
+        var adjuncts = await dbContext.Adjuncts.AsNoTracking().Where(a => a.AssetId == request.AssetId)
+            .ToListAsync(cancellationToken);
+        
+        return FetchEntityResult<IReadOnlyCollection<Adjunct>>.Success(adjuncts);
```
From ACs

> Returns 404 if asset doesn't exist, else 200 (collection may be empty if no adjuncts)

I don't think this will happen here? It looks like a 200 will be returned whether the asset exists or not?

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/GetAdjuncts.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace API.Features.Adjuncts.Requests;
+
+public class GetAdjuncts(AssetId assetId) : IRequest<FetchEntityResult<IReadOnlyCollection<Adjunct>>>, IPagedRequest
```
Small/consistency - should this be `GetAllAdjuncts`?

Other similar handlers use `GetAll{Thing}s` and `Get{Thing}` but this uses `Get{Thing}s` and `Get{Thing}`. E.g. `GetAllNamedQueries` and `GetNamedQuery`, `GetAllCustomers` and `GetCustomer`.

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/DeleteAdjunct.cs:
```diff
@@ -0,0 +1,36 @@
+            .Adjuncts
+            .Where(a => a.Id == request.Id && a.AssetId == request.AssetId)
+            .ExecuteDeleteAsync(cancellationToken);
+        await dbContext.SaveChangesAsync(cancellationToken);
```
`SaveChangesAsync()` is erroneous when using `ExecuteDeleteAsync()`, one of the benefits of `ExecuteDelete` is that there's no change tracking involved. 

See https://learn.microsoft.com/en-us/ef/core/saving/execute-insert-update-delete

> [ExecuteUpdate](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationalqueryableextensions.executeupdate) and [ExecuteDelete](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationalqueryableextensions.executedelete) are a way to save data to the database without using EF's traditional change tracking and [SaveChanges()](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.savechanges#microsoft-entityframeworkcore-dbcontext-savechanges) method.

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/CreateOrUpdateAdjunct.cs:
```diff
@@ -0,0 +1,73 @@
+    /// <summary>
+    /// Whether only creation is allowed (no update)
+    /// </summary>
+    public bool CreateOnly { get; set; } = createOnly;
```
These props can be `{ get; }` only

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/CreateOrUpdateAdjunct.cs:
```diff
@@ -0,0 +1,73 @@
+                a.Id == adjunct.Id && a.AssetId == adjunct.AssetId, cancellationToken);
+        }
+
+        if (dbAdjunct != null)
```
`Id` and `AssetId` can't change - is it worth omitting from this list?

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/CreateOrUpdateAdjunct.cs:
```diff
@@ -0,0 +1,73 @@
+        }
+        else
+        {
+            request.Adjunct.Created = DateTime.UtcNow;
```
For clarity, should we just use `adjunct` here? (because `var adjunct = request.Adjunct;`). Same with the return object.

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/CreateOrUpdateAdjunct.cs:
```diff
@@ -0,0 +1,73 @@
+        catch (DbUpdateException ex) when (ex.GetDatabaseError() is UniqueConstraintError)
+        {
+            return ModifyEntityResult<Adjunct>.Failure(
+                $"An adjunct called '{request.Adjunct.Id}' already exists",
```
`with id` rather than `called` ?

donaldgray commented on src/protagonist/API/Features/Adjuncts/Validation/HydraAdjunctValidator.cs:
```diff
@@ -0,0 +1,37 @@
+            .Must(a => Uri.IsWellFormedUriString(a, UriKind.Absolute))
+            .WithMessage("'externalId' is required and must be a well formed URI");
+        
+        RuleFor(a => a.IIIFLink).Must(a => Enum.IsDefined(typeof(IIIFLinkType), a!))
```
Small - is it worth moving this beside line 11 to keep validators for same fields together? When initially scanning code I thought we were only validating that `iiifLink` had a value.

donaldgray commented on src/protagonist/API/Features/Adjuncts/Validation/HydraAdjunctValidator.cs:
```diff
@@ -0,0 +1,37 @@
+        
+        RuleFor(a => a.Type)
+            .NotEmpty()
+            .WithMessage($"'@type' is required");
```
`$` can go

donaldgray commented on src/protagonist/DLCS.HydraModel/Adjunct.cs:
```diff
@@ -0,0 +1,79 @@
+    [JsonProperty(Order = 17, PropertyName = "externalId")]
+    public required string ExternalId { get; set; }
+    
+    [RdfProperty(Description = "Where this adjunct is hosted publically.",
```
`publically` -> `publicly`

donaldgray commented on src/protagonist/DLCS.Model/Assets/Adjunct.cs:
```diff
@@ -0,0 +1,77 @@
+    public string[]? Language { get; set; }
+    
+    /// <summary>
+    /// A fully-qualified URL external to the platform
```
worth adding something like "where the adjunct is hosted" or "where the adjunct can be accessed"?

donaldgray commented on src/protagonist/DLCS.Model/Assets/Adjunct.cs:
```diff
@@ -0,0 +1,77 @@
+    public required IIIFLinkType IIIFLink { get; set; }
+    
+    /// <summary>
+    /// The asset this adjunct is associated with
```
Splitting hairs but this is the Id of the Asset it's associated with, `public Asset Asset` is the actual asset the adjunct is associated with.

donaldgray commented on src/protagonist/DLCS.Repository/DlcsContext.cs:
```diff
@@ -697,6 +704,29 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             entity.HasKey(ba => new { ba.BatchId, ba.AssetId, });
         });
+        
+        modelBuilder.Entity<Adjunct>(entity =>
```
Where did these MaxLength values come from? Do we need them, if so should they differ? e.g. the longest `IIIFLink` value is `annotations`, which is 11 chars.

JackLewis-digirati replied:
it was based on other similar fields in the `context` class to not allow absolutely gigantic id's etc. - can likely be removed, but I felt they were useful to have as a basic stopgap, though the `iiifLink` property values are not needed - will remove that one

donaldgray commented on src/protagonist/API/Features/Adjuncts/Validation/HydraAdjunctValidator.cs:
```diff
@@ -0,0 +1,37 @@
+        
+        RuleFor(a => a.IIIFLink).Must(a => Enum.IsDefined(typeof(IIIFLinkType), a!))
+            .When(a => a.IIIFLink != null)
+            .WithMessage("Valid values for 'iiifLink' are 'SeeAlso', 'Annotations' and 'Rendering'");
```
Valid values are camel-case (`seeAlso`, `annotations` and `rendering`). 

We should enforce that only the correctly case is used be explicit (ie don't accept `SeeAlso`, it must be `seeAlso`).

donaldgray commented on src/protagonist/API/Converters/AdjunctConverter.cs:
```diff
@@ -0,0 +1,82 @@
+        
+        Debug.Assert(adjunctId != null, "adjunctId != null");
+
+        LanguageMap? label = null;
```
If `ToLanguageMap` was tolerant of nulls this block lines 45-49 could be removed

donaldgray commented on src/protagonist/API/Converters/AdjunctConverter.cs:
```diff
@@ -0,0 +1,82 @@
+public static class AdjunctConverter
+{
+    /// <summary>
+    /// This will create a DLCS.Model.Assets.Asset with the correct Id, Customer and Space.
```
XML comment from `Asset`

donaldgray commented on src/protagonist/API.Tests/Features/Adjuncts/Validation/HydraAdjunctValidatorTests.cs:
```diff
@@ -0,0 +1,144 @@
+    }
+    
+    [Fact]
+    public void ExternalId_Null()
```
Should this test null or empty?

donaldgray commented on src/protagonist/API.Tests/Features/Adjuncts/Validation/HydraAdjunctValidatorTests.cs:
```diff
@@ -0,0 +1,144 @@
+    }
+    
+    [Fact]
+    public void IiifLink_NotValid()
```
(tiny) `IIIFLink`, rather than `IiifLink` as `IIIF` is an acronym

donaldgray commented on src/protagonist/API.Tests/Integration/AdjunctTests.cs:
```diff
@@ -0,0 +1,424 @@
+    private readonly HttpClient httpClient;
+    private readonly DlcsContext dbContext;
+    
+    public AdjunctTests(StorageFixture storageFixture, ProtagonistAppFactory<Startup> factory)
```
This doesn't need `StorageFixture`, can just use `DlcsDatabaseFixture`. The former contains a `DlcsDatabaseFixture` but also a `LocalStackFixture` but that isn't required for these operations (it should marginally speed up tests/reduce overhead as LocalStack container won't need to run).

donaldgray commented on src/protagonist/API.Tests/Integration/AdjunctTests.cs:
```diff
@@ -0,0 +1,424 @@
+    }
+    
+    [Fact]
+    public async Task GetAdjunctS_RetrievesListOfAdjuncts()
```
Capital `S` here (same with others that test bulk fetch).

donaldgray commented on src/protagonist/API.Tests/Integration/AdjunctTests.cs:
```diff
@@ -0,0 +1,424 @@
+    }
+    
+    [Fact]
+    public async Task GetAdjunctS_RetrievesEmptyListOfAdjuncts_WhenNoAsset()
```
This should be a 404 rather than an empty list. From ACs

> Returns 404 if asset doesn't exist, else 200 (collection may be empty if no adjuncts)

donaldgray commented on src/protagonist/API.Tests/Integration/AdjunctTests.cs:
```diff
@@ -0,0 +1,424 @@
+    }
+    
+    [Fact]
+    public async Task DeleteAdjunct_NothingHappens_WhenDoesNotExist()
```
Small - is `DeleteAdjunct_Returns404` better than `DeleteAdjunct_NothingHappens`?

donaldgray reviewed (changes requested):
Overall LGTM but a couple of little tweaks required - most comments are small naming etc.

JackLewis-digirati committed:
Code review fixes

- Various naming/grammatical changes
- Make iiifLink camel case
- Remove unneeded paging values
- Make get all return 404 if asset doesn't exist
- Make ToLanguageMap nullable
- Remove unneeded Sets
- Remove DB length limit on iiif link
- Move test class to db fixture

donaldgray commented on src/protagonist/API/Converters/AdjunctConverter.cs:
```diff
@@ -0,0 +1,82 @@
+        {
+            Type = adjunct.Type,
+            MediaType = adjunct.MediaType,
+            IIIFLink = adjunct.IIIFLink.ToString().ToCamelCase(true),
```
Rather than `.ToCamelCase()` here, is it worth using `DescriptionAttribute` to explicitly specify the value we want the enum represented as? There are helpers in `EnumX` for reading

donaldgray replied:
Update after submitting review - noticed that the mapping in dbContext is using `EnumX.GetDescription` so that would keep the value stored in DB consistent with value returned from API

donaldgray commented on src/protagonist/API/Converters/AdjunctConverter.cs:
```diff
@@ -0,0 +1,82 @@
+        
+        Debug.Assert(adjunctId != null, "adjunctId != null");
+
+        LanguageMap? label = null;
```
Why split declaration from assignment here? Also, don't think this var is needed at all?

donaldgray commented on src/protagonist/API/Converters/LanguageMapX.cs:
```diff
@@ -0,0 +1,24 @@
+public static class LanguageMapX
+{
+    public static LanguageMap? ToLanguageMap(this Dictionary<string, List<string>>? dictionaryToConvert)
+    {
```
Feels a lot cleaner to start this with `if (dictionaryToConvert == null) return null;` and return early. 

Can then just assign `var lm = new LanguageMap();` and remove a level of nesting

donaldgray commented on src/protagonist/API/Features/Adjuncts/Requests/GetAllAdjuncts.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    public async Task<FetchEntityResult<IReadOnlyCollection<Adjunct>>> Handle(GetAllAdjuncts request, CancellationToken cancellationToken)
+    {
+        var assetExists = dbContext.Images.Any(i => i.Id == request.AssetId);
```
Sub-optimal to make 2 requests to DB to fulfil this request. I'd imagine in real use it'll be rare to call this when the image doesn't exist.

donaldgray commented on src/protagonist/API/Features/Adjuncts/Validation/HydraAdjunctValidator.cs:
```diff
@@ -0,0 +1,42 @@
+            .WithMessage("'@type' is required");
+    }
+
+    private readonly List<string> validIiifLinkTypes =
```
`Iiif` (also, if using `[Description]` for values then this would need to change too)

JackLewis-digirati committed:
Code review changes

- use GetDescription over ToCamelCase
- Spelling changes
- Explicit ordering by id
- Simplify null check

donaldgray committed:
Tweak GetAllAdjuncts query

Only return Images Id to determine existance and order in db
